README.md
# Golang Chess

This project is bare minimum for building a chess site with golang.
It uses javascript chess libraries: Chessboard.js, Chess.js, and the Garbo Chess Engine.

Please visit the project site: http://chessbook.org

## Installation

TODO

## Features

* Guest can play againts computer (GarboChess.js, 800ms)
* Guest can play againts another random guest
* Guest can watch random game

## Not Supported

Below features are not and will not supported. You must implement them.
* Chess game by time
* Draw detection by repetition or by 50 moves
* Registered users
* Saving chess games
* ELO or other rank system
* Play againts diferrent levels of computer
* Pawn promotion will be only queen

# Working with GarboChess.js

To work with GarboChess.js I need to understand some other programmers code. One or two projects use Gary Linscott's 
main codes from boardui.js. I figured it out how to include only garbochess.js and not include boardui.js (it is a board ui, and we dont need it). I will explain my way. You need to modify for your needs.

First we created a worker:
```js
var g_garbo = new Worker("/js/garbochess.js");
```
To check messages from worker, we need an onmessage function:
```js
g_garbo.onmessage = function (e) {
   ...
}
```
And then send messages to worker:
```js
g_garbo.postMessage("position " + game.fen());
g_garbo.postMessage("search 800");
```
First line sets up a position, second line says "Solve it in 800 ms."

And I check messages from worker like this:
```js
g_garbo.onmessage = function (e) {
    move = { from: e.data[0]+e.data[1], to: e.data[2]+e.data[3] };
    if (e.data[4]) {
        move.promotion = e.data[4];
    }
    g_garbo.terminate();
    game.move(move);
    board.position(game.fen());
    socket.emit("move", {move: move, fen: game.fen()});
}
```
CAUTION: Since I modified GarboChess.js (latest "Test Harness" part) I only get the best move as message.
But original GarboChess.js send more messages if you like it analyze the position. Then you may need something like that:
```js
g_garbo.onmessage = function (e) {
    if (e.data.match("^pv") == "pv") {
    } else if (e.data.match("^sam") == "sam") {
    } else if (e.data.match("^message") == "message") {
    } else {
        move = { from: e.data[0]+e.data[1], to: e.data[2]+e.data[3] };
        if (e.data[4]) {
            move.promotion = e.data[4];
        }
        g_garbo.terminate();
        game.move(move);
        board.position(game.fen());
        socket.emit("move", {move: move, fen: game.fen()});
    }        
}
```


history.txt
Golang Chess
Fahri Aydos (aydos.com|.net|.org|.io)

2015 08 04 - v0.2
* mesajlaşma yenilendi, json kullanılmıyor artık

2015 07 24 - v0.1
* local'de git kullanılmaya başlandı

2015 07 10 - v0.1
* github'a kondu

2015 06 26 - v0.0
* isim "golang chess" olaraka değiştirildi

2015 06 19 - v0.0
* golang ile ilk kod, ismi "Go ChessBook"



golangchess.go

package main

import (
    "fmt"
    "log"
    "math/rand"
    "net/http"
    "time"
    "strings"
    ws "github.com/gorilla/websocket"
)

type user struct {
    name string
    status string // idle, wait, play, watch
    game string
    color string
    opponent *ws.Conn
}

type game struct {
    id string
    white *ws.Conn
    black *ws.Conn
    fen string // "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
    watchers map[*ws.Conn]bool
}

// http://egypt.silverkeytech.com/blogs/third-world/2013/8/you-cannot-assign-a-struct-field-off-map-directly
var users map[*ws.Conn]*user = make(map[*ws.Conn]*user)
var games map[string]*game = make(map[string]*game)
var wait *ws.Conn = nil

func main() {
    log.Println("Golang-Chess'e Hoşgeldiniz....")
	rand.Seed(time.Now().UnixNano())
	http.Handle("/", http.FileServer(http.Dir("./client")))
	http.HandleFunc("/ws", wsHandler)
    log.Fatal(http.ListenAndServe(":8080", nil))
}

// ******************************************************************
// Gorilla WebSocket

func wsHandler(w http.ResponseWriter, r *http.Request) {
    conn, err := ws.Upgrade(w, r, nil, 1024, 1024);

    if _, ok := err.(ws.HandshakeError); ok {
        http.Error(w, "Not a websocket handshake", 400)
        return
    } else if err != nil {
        log.Println(err)
        return
    }

    defer conn.Close()

    users[conn] = newUser()

    for {

        // read message
        _, p, err := conn.ReadMessage();
        if err != nil {
            log.Println(err)
            deleteUser(conn)
            broadcastCounts()
            return
        }

        // split message into array
        m := strings.Split(string(p), ",")

        if m[0] == "init" {
            sendUserInfo(conn, "user")
            broadcastCounts();
        }

        if m[0] == "pc" { // play computer
            id := newGame(conn, nil);
            broadcastCounts();
            broadcastGameInfo(id);
        }

        if m[0] == "ph" { // play human
            if wait == nil {
                wait = conn;
                users[conn].status = "wait";
                sendUserInfo(conn, "wait")
            } else {
                opponent := wait
                wait = nil
                id := newGame(opponent, conn)
                broadcastCounts()
                broadcastGameInfo(id)
            }
        }

        if m[0] == "cg" { // cancel watch game
            cancelWatch(conn)
        }

        if m[0] == "cw" { // cancel wait
            if (wait == conn) {
                wait = nil
            }
            users[conn].status = "idle"
            sendUserInfo(conn, "user")
        }

        if m[0] == "wg" { // watch game
            id := gameRand()
            if id != "" {
                games[id].watchers[conn] = true
                users[conn].status = "watch"
                users[conn].game = id;
                sendUserInfo(conn, "user")
                p := []byte(fmt.Sprintf("wg,%s", games[id].fen))
                if err := conn.WriteMessage(ws.TextMessage, p); err != nil {
                    log.Println(err)
                }
                broadcastGameInfo(id);
            }
        }

        if m[0] == "fg" { // finish game
            id := users[conn].game
            finishGame(id, m[1])
            broadcastCounts()
        }

        if m[0] == "mv" { // move
            // m[1] : move
            // m[2] : fen
            p := []byte(fmt.Sprintf("mv,%s", m[1]))
            var opponent = users[conn].opponent;
            if (opponent != nil) {
                if err := opponent.WriteMessage(ws.TextMessage, p); err != nil {
                    log.Println(err)
                }
            }
            var id = users[conn].game;
            games[id].fen = m[2]
            for c := range games[id].watchers {
                if err := c.WriteMessage(ws.TextMessage, p); err != nil {
                    log.Println(err)
                }
            }
        }

    }
}

func broadcastCounts() {
    p := []byte(fmt.Sprintf("c,%d,%d", len(users), len(games)))
    for conn := range users {
        if err := conn.WriteMessage(ws.TextMessage, p); err != nil {
            log.Println(err)
        }
    }
}

func broadcastGameInfo(id string) {
    whitename := users[games[id].white].name
    blackname := ""
    if (games[id].black == nil) {
        blackname = "computer"
    } else {
        blackname = users[games[id].black].name
    }
    p := []byte(fmt.Sprintf("gi,%s,%s,%d", whitename, blackname, len(games[id].watchers)))
    // send to white
    if err := games[id].white.WriteMessage(ws.TextMessage, p); err != nil {
        log.Println(err)
    }
    // send to black
    if (games[id].black != nil) {
        if err := games[id].black.WriteMessage(ws.TextMessage, p); err != nil {
            log.Println(err)
        }
    }
    // send to watchers
    for c := range games[id].watchers {
        if err := c.WriteMessage(ws.TextMessage, p); err != nil {
            log.Println(err)
        }
    }
}

func sendUserInfo(conn *ws.Conn, t string) {
    opponent := ""
    if users[conn].game != "" {
        if users[conn].opponent == nil {
            opponent = "computer"
        } else {
            opponent = users[users[conn].opponent].name
        }
    }
    p := []byte(fmt.Sprintf("%s,%s,%s,%s,%s", t, users[conn].name, users[conn].status, users[conn].color, opponent))
    if err := conn.WriteMessage(ws.TextMessage, p); err != nil {
        log.Println(err)
        return
    }
}

// generate new user add to users list
func newUser() *user {
   // new user, name may not be unique
    var newuser = user {
        name : hexRand(),
        status : "idle",    // idle, wait, play, watch
        game : "",
        color : "",
        opponent : nil,
    }
    return &newuser;
}

func deleteUser(conn *ws.Conn) {
    id := users[conn].game;
    status := users[conn].status
    if (status == "play") {
        why := "6"                          // white disconnected
        if users[conn].color == "black" {   // black disconnected
            why = "7"
        }
        finishGame(id, why);
    } else if (status == "watch") {
        if _, ok := games[id]; ok {
            delete(games[id].watchers, conn)
            broadcastGameInfo(id);
        }
    }
    if (wait == conn) {
        wait = nil;
    }
    delete(users, conn);
}

func newGame(white, black *ws.Conn) string {
    // generate non exists game
    id := hexRand();
    _, ok := games[id]
    for ok {
        id = hexRand();
        _, ok = games[id]
    }
    // new game
    var newgame = game{
        id : id,
        white : white,
        black : black,
        fen : "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
        watchers : map[*ws.Conn]bool{},
    }
    games[id] = new(game)
    *games[id] = newgame
    users[white].status = "play";
    users[white].game = newgame.id;
    users[white].color = "white";
    users[white].opponent = black;
    sendUserInfo(white, "sg")
    if (black != nil) {
        users[black].status = "play"
        users[black].game = newgame.id;
        users[black].color = "black"
        users[black].opponent = white
        sendUserInfo(black, "sg")
    }
    return id;
}

func cancelWatch(conn *ws.Conn) {
    id := users[conn].game
    if _, ok := games[id]; ok {
        delete(games[id].watchers, conn)
        broadcastGameInfo(id)
    }
    users[conn].status = "idle"
    users[conn].game = ""
    users[conn].color = ""
    users[conn].opponent = nil
    sendUserInfo(conn, "user")
}


func finishGame(id, why string) {
    if _, ok := games[id]; !ok {
        return
    }
    // save users info
    white := games[id].white
    black := games[id].black
    watchers := games[id].watchers

    // delete game
    delete(games, id)

    p := []byte(fmt.Sprintf("fg,%s", why))

    // do user staff
    if (why != "6") { // if white not disconnected
        users[white].status = "idle"
        users[white].game = ""
        users[white].color = ""
        users[white].opponent = nil
        sendUserInfo(white, "fu")
        if err := white.WriteMessage(ws.TextMessage, p); err != nil {
            log.Println(err)
            return
        }
    }
    if (black != nil && why != "7") { // if black not disconnected
        users[black].status = "idle"
        users[black].game = ""
        users[black].color = ""
        users[black].opponent = nil
        sendUserInfo(black, "fu")
        if err := black.WriteMessage(ws.TextMessage, p); err != nil {
            log.Println(err)
            return
        }
    }
    for c := range watchers {
        users[c].status = "idle"
        users[c].game = ""
        users[c].color = ""
        users[c].opponent = nil
        sendUserInfo(c, "fu")
        if err := c.WriteMessage(ws.TextMessage, p); err != nil {
            log.Println(err)
            return
        }
    }
}


// ******************************************************************
// Random generators

func Rand(min, max int) int {
	return rand.Intn(max - min + 1) + min;
}

func hexRand() string {
    var hex = make([]byte, 8)
    var chrs = "0123456789ABCDEF"
    for i:=0; i<8; i++ {
        hex[i] = chrs[Rand(0,15)];
    }
    return string(hex);
}

func gameRand() string {
    if len(games) <= 0 {
        return ""
    }
    ind := Rand(1, len(games))
    i := 1
    for g := range games {
        if i == ind {
            return g
        }
        i++
    }
    return ""
}




client/js/golangchess.js

// chess game variables
var ws, user, game, board, cfgplay, cfgwatch;

// jquery variables
var $username, $usercnt, $gamecnt, $gameinfo;
var $hider, $status, $askplay;
var $turnup, $turndown;
var $playerup, $playerdown;
var $message, $cancel;
var $watchgames;
var $watcherinfo;
var $aydos, $about;

// main function
$(document).ready(function() {
    user = {};
    //game = new Chess();
    cfgplay = {
        draggable: true,
        position: "start",
        onDragStart: onDragStart,
        dropOffBoard: "snapback",
        onDrop: onDrop,
        onSnapEnd: onSnapEnd,
        pieceTheme: "/img/alpha/{piece}.png"
    };
    cfgwatch = {
        position: "start",
        pieceTheme: "/img/alpha/{piece}.png"
    };
    board = new ChessBoard("board", cfgplay);

    $username = $("#username");
    $usercnt = $("#usercnt");
    $gamecnt = $("#gamecnt");
    $gameinfo = $("#gameinfo");
    $hider = $("#hider");
    $status = $("#status");
    $askplay = $("#askplay");
    $turnup = $("#turnup");
    $turndown = $("#turndown");
    $watchgames = $("#watchgames");
    $playerup = $("#playerup");
    $playerdown = $("#playerdown");
    $message = $("#message");
    $cancel = $("#cancel");
    $watcherinfo = $("#watcherinfo");
    $aydos = $("#aydos");
    $about = $("#about");

    $(window).bind("resize", initWindow);
    $(document).bind("keydown", keyDown);

    initWindow();
    clickFunctions();

    if (window["WebSocket"]) {
        ws = new WebSocket("ws://" + window.location.host + "/ws");
        ws.onopen = function() {
            ws.send("init");
        }
        ws.onmessage = function(e) {
            getdata(e.data)
        }
        ws.onclose = function(evt) {
        }
    } else {
        console.log("Your browser does not support WebSockets.");
    }

    askplay();
});

function getdata(data) {

    var m = data.split(",");

    if (m[0] == "user") {           // user info
        user.name = m[1];
        user.status = m[2];
        user.color = m[3];
        user.opponent = m[4];
        $username.html(m[1]);
        if (m[2] == "idle") {
            askplay();
        }
    }

    else if (m[0] == "c") {         // counts
        $usercnt.html(m[1]);
        $gamecnt.html(m[2]);
        if (m[2] == 0) {
            $watchgames.prop("disabled", true);
        } else {
            $watchgames.prop("disabled", false);
        }
    }

    else if (m[0] == "sg") {        // start game
        $hider.fadeOut("fast");
        $status.fadeOut("fast");
        user.name = m[1];
        user.status = m[2];
        user.color = m[3];
        user.opponent = m[4];
        game = new Chess();
        board = new ChessBoard("board", cfgplay);
        board.orientation(user.color);
        if(user.color == "white") {
            $turnup.removeClass("turnwhite").addClass("turnblack");
            $turndown.removeClass("turnblack").addClass("turnwhite");
        } else {
            $turnup.removeClass("turnblack").addClass("turnwhite");
            $turndown.removeClass("turnwhite").addClass("turnblack");
        }
        showTurn();
    }

    else if (m[0] == "wg") {
        $hider.fadeOut("fast");
        $status.fadeOut("fast");
        game = new Chess(m[1]);
        board = new ChessBoard("board", cfgwatch);
        board.position(m[1]);
        board.orientation("white");
        $turnup.removeClass("turnwhite").addClass("turnblack");
        $turndown.removeClass("turnblack").addClass("turnwhite");
        showTurn();
    }

    else if (m[0] == "fu") {  // finish user
        user.name = m[1];
        user.status = m[2];
        user.color = m[3];
        user.opponent = m[4];
    }

    else if (m[0] == "fg") { // finish game
        status = "Game finished."
        if (m[1] == 1) {
            status = "Draw"
        } else if (m[1] == 2) {
            status = "White won"
        } else if (m[1] == 3) {
            status = "Black won"
        } else if (m[1] == 4) {
            status = "White won. Black resigned."
        } else if (m[1] == 5) {
            status = "Black won. White resigned."
        } else if (m[1] == 6) {
            status = "White won. Black disconnected."
        } else if (m[1] == 7) {
            status = "Black won. White disconnected."
        }
        showStatus(status, "OK");
    }

    else if (m[0] == "wait") {
        user.name = m[1];
        user.status = m[2];
        user.color = m[3];
        user.opponent = m[4];
        showStatus("Waiting for opponent", "Cancel");
    }

    else if (m[0] == "gi") {
        if (board.orientation() === "white") {
            $playerup.html(m[2]);
            $playerdown.html(m[1]);
        } else {
            $playerup.html(m[1]);
            $playerdown.html(m[2]);
        }
        if (m[3] > 0) {
            $watcherinfo.html("Watcher this game:<br /><span id='watchercnt'>"+ m[3] + "</span>");
        } else {
            $watcherinfo.html("");
        }
    }

    else if (m[0] == "mv") {
        move = { from: m[1][0]+m[1][1], to: m[1][2]+m[1][3] };
        if (m[1][4]) {
            move.promotion = m[1][4];
        }
        game.move(move);
        board.position(game.fen());
        showTurn();
    }

}

function initWindow() {
    var w = document.body.clientWidth;
    var h = document.body.clientHeight;
    
    var gh = h - $("#menu").outerHeight() - 24;
    var gw = gh + 400;
    $("#game").css({width: gw, height: gh});
    $("#board").css({width: gh, height: gh});
    $("#infoup").css({height: gh});
    $hider.css({width: 2*w, height: 2*h, "margin-left": (0-w), "margin-top": (0-h)});

    board.resize();
}

function keyDown(e) {
    if (e.keyCode == 27) {      // esc
        e.preventDefault();
        if (user.status === "play") {
            if ($status.is(':visible')) { // cancel resign
                $hider.fadeOut("fast");
                $status.fadeOut("fast");
            } else {                           // ask resign
                showStatus("Do you resign the game?", "Yes");
            }
        } else if (user.status === "watch") {  // leave watching game
            ws.send("cw");
        }
    }
}

function showStatus(message, button) {
    $hider.fadeIn("fast");
    $status.fadeIn("fast");
    $message.html(message);
    $cancel.html(button);
}

function showTurn() {
    var turn = game.turn() === "w" ? "white" : "black";
    if (turn === board.orientation()) {
        $turnup.hide();
        $turndown.show();
    } else {
        $turnup.show();
        $turndown.hide();
    }
}

function clickFunctions() {

    $("#btnabout").click(function() {
        $hider.fadeIn("fast");
        $about.fadeIn("fast");
    });

    $("#btnaydos").click(function() {
        $hider.fadeIn("fast");
        $aydos.fadeIn("fast");
    });

    $("#closeabout").click(function() {
        $hider.fadeOut("fast");
        $about.fadeOut("fast");
    });

    $("#closeaydos").click(function() {
        $hider.fadeOut("fast");
        $aydos.fadeOut("fast");
    });

    askplay = function() {
        $hider.fadeIn("fast");
        $askplay.fadeIn("fast");
    };

    $("#playhuman").click(function() {
        $hider.fadeOut("fast");
        $askplay.fadeOut("fast");
        ws.send("ph");
    });

    $("#playcomputer").click(function() {
        $hider.fadeOut("fast");
        $askplay.fadeOut("fast");
        ws.send("pc");
    });

    $("#watchgames").click(function() {
        $hider.fadeOut("fast");
        $askplay.fadeOut("fast");
        ws.send("wg");
    });

    $cancel.click(function() {
        $status.fadeOut("fast");
        $askplay.fadeIn("fast");
        if ($cancel.html() === "Cancel") {
            ws.send("cw");
        }
        if ($cancel.html() === "OK") {
            ;
        }
        if ($cancel.html() === "Yes") {
            if (user.color == "white") {
                status = 4; // white resigned
            } else {
                status = 5; // black resigned
            }
            ws.send("fg,"+status);
        }
    });

}


// ====================================================
// Integration between chessboard.js and chess.js

// do not pick up pieces if the game is over
// only pick up pieces for the side to move
var onDragStart = function(source, piece, position, orientation) {
    if (game.game_over() === true || (game.turn() != piece[0]) ) {
        return false;
    }
    if (game.turn() === "w" && user.color === "black") {
        return false;
    }
    if (game.turn() === "b" && user.color === "white") {
        return false;
    }
};

var onDrop = function(source, target) {
    // see if the move is legal
    var move = game.move({
        from: source,
        to: target,
        promotion: "q"
    });

    move = source + target + "q";
    // illegal move
    if (move === null) return "snapback";

    // make the move
    ws.send("mv," + move + "," + game.fen());

    // if play againts computer
    if (user.opponent === "computer") {
        makeComputerMove();
    }

    updateStatus();
};

// update the board position after the piece snap 
// for castling, en passant, pawn promotion
var onSnapEnd = function() {
    board.position(game.fen());
    showTurn();
};

var updateStatus = function() {
    var status = 0;
    /*var moveColor = 'White';
    if (game.turn() === 'b') {
        moveColor = 'Black';
    }*/
    // checkmate?
    if (game.in_checkmate() === true) {
        if (game.turn() === 'b') {
            status = 2; // white won
        } else {
            status = 3; // black won
        }
        //status = 'Game over, ' + moveColor + ' is in checkmate.';
        ws.send("fg,"+status);
    }
    // draw?
    else if (game.in_draw() === true) {
        status = 1;  // draw
        ws.send("fg,"+status);
    }
    // game still on
    else {
        /*status = moveColor + ' to move';
        // check?
        if (game.in_check() === true) {
            status += ', ' + moveColor + ' is in check';
        }*/
    }
};

var makeComputerMove = function() {
    var possibleMoves = game.moves();

    // game over
    if (possibleMoves.length === 0) return;

    /*
    // Random Move
    var randomIndex = Math.floor(Math.random() * possibleMoves.length);
    game.move(possibleMoves[randomIndex]);
    board.position(game.fen());
    ws.send("mv," + possibleMoves[randomIndex] + "," + fen);
    */

    // GarboChess
    var g_garbo = new Worker("/js/garbochess.min.js");
    g_garbo.onmessage = function (e) {
        move = { from: e.data[0]+e.data[1], to: e.data[2]+e.data[3] };
        if (e.data[4]) {
            move.promotion = e.data[4];
        }
        g_garbo.terminate();
        game.move(move);
        board.position(game.fen());
        showTurn();
        ws.send("mv," + e.data + "," + game.fen());
        updateStatus();
    }
    g_garbo.postMessage("position " + game.fen());
    g_garbo.postMessage("search 800");

};

